### PR TITLE
[SPARK-40487][SQL] Make defaultJoin in BroadcastNestedLoopJoinExec running in parallel

### PR DESCRIPTION
**Pull request by xingchaozh**
### What changes were proposed in this pull request?

Currently, the defaultJoin method in BroadcastNestedLoopJoinExec collects notMatchedBroadcastRows firstly, then collects matchedStreamRows. The two steps could run in parallel instead of serial.

### Why are the changes needed?
Make defaultJoin in BroadcastNestedLoopJoinExec running in parallel.

### Does this PR introduce _any_ user-facing change?
NO.

### How was this patch tested?
UT.

### CODE REVIEW

**Comment by wangyum** [comment]
@cloud-fan @c21

File: sql/core/src/main/scala/org/apache/spark/sql/execution/joins/BroadcastNestedLoopJoinExec.scala [hunk 1]
```diff
@@ -286,21 +286,26 @@ case class BroadcastNestedLoopJoinExec(
...
-      joinedRow.withLeft(nulls)
-      var i = 0
-      val buildRows = relation.value
-      while (i < buildRows.length) {
-        if (!matchedBroadcastRows.get(i)) {
-          buf += joinedRow.withRight(buildRows(i)).copy()
+    def notMatchedBroadcastRows: RDD[InternalRow] = {
+      getMatchedBroadcastRowsBitSetRDD(streamRdd, relation).
+        repartition(1).
+        mapPartitions(iter => {
```
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/joins/BroadcastNestedLoopJoinExec.scala:292]:
```suggestion
      getMatchedBroadcastRowsBitSetRDD(streamRdd, relation)
        .repartition(1)
        .mapPartitions(iter => {
```

**Review by cloud-fan — APPROVED** [review]

**Comment by cloud-fan** [comment]
makes sense, also cc @maryannxue @somani

File: sql/core/src/main/scala/org/apache/spark/sql/execution/joins/BroadcastNestedLoopJoinExec.scala (hunk 1 above)
Comment (xingchaozh) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/joins/BroadcastNestedLoopJoinExec.scala]: Thanks for the review! just fixed the code style

File: sql/core/src/main/scala/org/apache/spark/sql/execution/joins/BroadcastNestedLoopJoinExec.scala [hunk 2]
```diff
@@ -342,6 +347,13 @@ case class BroadcastNestedLoopJoinExec(
   private def getMatchedBroadcastRowsBitSet(
       streamRdd: RDD[InternalRow],
       relation: Broadcast[Array[InternalRow]]): BitSet = {
+    getMatchedBroadcastRowsBitSetRDD(streamRdd, relation).
+      fold(new BitSet(relation.value.length))(_ | _)
```
Comment (wangyum) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/joins/BroadcastNestedLoopJoinExec.scala:351]:
```suggestion
    getMatchedBroadcastRowsBitSetRDD(streamRdd, relation)
      .fold(new BitSet(relation.value.length))(_ | _)
```

File: sql/core/src/test/scala/org/apache/spark/sql/JoinSuite.scala [hunk 3]
```diff
@@ -1440,4 +1440,25 @@ class JoinSuite extends QueryTest with SharedSparkSession with AdaptiveSparkPlan
...
+
+  test("SPARK-40487: Make defaultJoin in BroadcastNestedLoopJoinExec running in parallel") {
+    withTable("t1", "t2") {
+      spark.range(5, 15).toDF("k").write.saveAsTable("t1")
+      spark.range(4, 8).toDF("k").write.saveAsTable("t2")
+
+      val queryBuildLeft =
+        s"""
+           |SELECT /*+ BROADCAST(t1) */ *  FROM t1 LEFT JOIN t2 on t1.k < t2.k
+       """.stripMargin
```
Comment (wangyum) [inline sql/core/src/test/scala/org/apache/spark/sql/JoinSuite.scala:1452]:
```suggestion
val queryBuildLeft = "SELECT /*+ BROADCAST(t1) */ * FROM t1 LEFT JOIN t2 on t1.k < t2.k"
```

File: sql/core/src/test/scala/org/apache/spark/sql/JoinSuite.scala [hunk 4]
```diff
@@ -1440,4 +1440,25 @@ class JoinSuite extends QueryTest with SharedSparkSession with AdaptiveSparkPlan
...
+      val result1 = sql(queryBuildLeft)
+
+      val queryBuildRight =
+        s"""
+           |SELECT /*+ BROADCAST(t2) */ *  FROM t1 LEFT JOIN t2 on t1.k < t2.k
+       """.stripMargin
```
Comment (wangyum) [inline sql/core/src/test/scala/org/apache/spark/sql/JoinSuite.scala:1458]:
```suggestion
val queryBuildRight = "SELECT /*+ BROADCAST(t2) */ * FROM t1 LEFT JOIN t2 ON t1.k < t2.k"
```

**Review by wangyum — APPROVED** [review]

File: sql/core/src/main/scala/org/apache/spark/sql/execution/joins/BroadcastNestedLoopJoinExec.scala [hunk 5]
```diff
@@ -286,21 +286,26 @@ case class BroadcastNestedLoopJoinExec(
...
-      val buildRows = relation.value
-      while (i < buildRows.length) {
-        if (!matchedBroadcastRows.get(i)) {
-          buf += joinedRow.withRight(buildRows(i)).copy()
+    def notMatchedBroadcastRows: RDD[InternalRow] = {
+      getMatchedBroadcastRowsBitSetRDD(streamRdd, relation)
+        .repartition(1)
+        .mapPartitions { iter =>
+          Seq(iter.fold(new BitSet(relation.value.length))(_ | _)).toIterator
+        }.flatMap(matchedBroadcastRows => {
```
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/joins/BroadcastNestedLoopJoinExec.scala:294]:
```suggestion
        }.flatMap {matchedBroadcastRows =>
```

File: sql/core/src/main/scala/org/apache/spark/sql/execution/joins/BroadcastNestedLoopJoinExec.scala (hunk 2 above)
Comment (xingchaozh) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/joins/BroadcastNestedLoopJoinExec.scala]: Sure, fixed

File: sql/core/src/test/scala/org/apache/spark/sql/JoinSuite.scala (hunk 3 above)
Comment (xingchaozh) [reply sql/core/src/test/scala/org/apache/spark/sql/JoinSuite.scala]: Sure, fixed

File: sql/core/src/test/scala/org/apache/spark/sql/JoinSuite.scala (hunk 4 above)
Comment (xingchaozh) [reply sql/core/src/test/scala/org/apache/spark/sql/JoinSuite.scala]: Sure, fixed

File: sql/core/src/main/scala/org/apache/spark/sql/execution/joins/BroadcastNestedLoopJoinExec.scala (hunk 5 above)
Comment (xingchaozh) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/joins/BroadcastNestedLoopJoinExec.scala]: Sure, fixed

**Comment by wangyum** [comment]
Merged to master.

**Pull request closed without merging**